pkg/cache/simple.go

```diff
@@ -0,0 +1,164 @@
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+
+package cache
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/envoyproxy/go-control-plane/api"
+	"github.com/golang/protobuf/proto"
+)
+
+// SimpleCache is a snapshot-based cache that maintains a single version per
+// xDS response, node group tuple, with no canary updates.
+type SimpleCache struct {
+	// responses are cached resources
+	responses map[Key]map[ResponseType][]proto.Message
+
+	// versions must have the same key set as responses
+	versions map[Key]map[ResponseType]int
+
+	// watches keeps track of open watches
+	watches map[Key]map[ResponseType]map[int64]Watch
+
+	// callback requests missing responses
+	callback func(Key, ResponseType)
+
+	// watchCount is the ID generator for watches
+	watchCount int64
+
+	// groups is the hashing function for proxy nodes
+	groups NodeGroup
+
+	mu sync.Mutex
+}
+
+// MakeSimpleCache initializes a simple cache.
+// callback function is called on every new cache key and response type if there is no response available.
+// callback is executed in a go-routine.
+func MakeSimpleCache(groups NodeGroup, callback func(Key, ResponseType)) Cache {
+	return &SimpleCache{
+		responses:  make(map[Key]map[ResponseType][]proto.Message),
+		versions:   make(map[Key]map[ResponseType]int),
+		watches:    make(map[Key]map[ResponseType]map[int64]Watch),
+		callback:   callback,
+		watchCount: 0,
+		groups:     groups,
+	}
+}
+
+func (cache *SimpleCache) SetResource(group Key, typ ResponseType, resources []proto.Message) {
+	cache.mu.Lock()
+	defer cache.mu.Unlock()
+
+	// update the existing entry
+	if _, ok := cache.responses[group]; !ok {
+		cache.responses[group] = make(map[ResponseType][]proto.Message)
+		cache.versions[group] = make(map[ResponseType]int)
+	}
+
+	cache.responses[group][typ] = resources
+
+	// increment version
+	cache.versions[group][typ]++
+
+	// trigger watches
+	if _, ok := cache.watches[group]; ok {
+		if _, ok := cache.watches[group][typ]; ok {
+			// signal the update (channels have buffer size 1)
+			for _, watch := range cache.watches[group][typ] {
+				watch.Value <- Response{
+					Version:   fmt.Sprintf("%d", cache.versions[group][typ]),
+					Resources: resources,
+				}
+				// remove clean-up as the watch is discarded immediately
+				watch.stop = nil
+			}
+			// discard all watches; the client must request a new watch to receive updates after ACK/NACK
+			cache.watches[group][typ] = make(map[int64]Watch)
```

Comment:
It would be good to have stats, as we have in Istio. At some point, this will enable us to profile as well.
This might be too much to ask. So feel free to defer it for follow up PR.